File: app/javascript/toast_stream.js
import { Turbo } from "@hotwired/turbo-rails"

// The server's way into the toast host. A controller answering with
//
//   render turbo_stream: turbo_stream.toast(t("flash.saved"), kind: :success)
//
// sends <turbo-stream action="toast" kind="success" target="toasts"> with the
// message as its template's text, and this is what that action does with it.
//
// It raises the same `toast:show` event a Stimulus controller raises rather
// than writing a row itself. That is the whole point: the row's markup lives in
// one <template> in shared/_toasts, and its reveal, its clock and its removal
// live in one controller. A stream that appended finished markup would need a
// second copy of both, and the two would drift.
//
// Everything but the message rides as an attribute, and the message is read as
// text and written as text — see the escaping in
// config/initializers/toast_stream.rb — so a message carrying a student's own
// words cannot carry markup with them.
//
// Outside controllers/ for the same reason frame_recovery.js is: `pin_all_from`
// would otherwise register it as a Stimulus controller, and this is one
// registration, not a behaviour attached to an element.
Turbo.StreamActions.toast = function () {
  const message = this.templateContent.textContent.trim()
  if (!message) return

  const read = name => this.getAttribute(name) || undefined
  const label = read("action-label")
  const href = read("action-href")

  window.dispatchEvent(new CustomEvent("toast:show", {
    detail: {
      message,
      kind: read("kind") || "info",
      title: read("title"),
      // Absent means "use the controller's default"; "0" means "stay".
      duration: this.hasAttribute("duration") ? Number(this.getAttribute("duration")) : undefined,
      action: label && href ? { label, href, method: read("action-method") } : undefined
    }
  }))
}

File: config/initializers/toast_stream.rb
# `turbo_stream.toast "Saved", kind: :success` — the server half of the toast
# host in shared/_toasts. See app/javascript/toast_stream.js for what the
# browser does with the tag this builds.
#
# Registered through Turbo's own extension point rather than by reopening the
# class, so it survives the gem loading whenever it likes. The block is
# instance_eval'd on the tag builder, which is why the kinds are a local rather
# than a constant: a constant written here would land on Object.
ActiveSupport.on_load :turbo_streams_tag_builder do
  # The message goes into the template as *escaped text*, not as markup.
  # `turbo_stream_action_tag` marks whatever it is handed as html_safe before
  # wrapping it, so escaping here is what stops a message built from a student's
  # own input carrying markup into the page.
  #
  # Everything else rides as an attribute, which `tag` escapes:
  #
  #   kind      one of info, success, warning, error. An unknown one raises:
  #             it would render as plain info and read as a styling bug rather
  #             than as the typo it is.
  #   title     an optional bold line above the message.
  #   duration  milliseconds. 0 keeps the toast until it is dismissed. Omitted
  #             leaves the controller's default alone — which is why it is not
  #             defaulted here.
  #   action    { label:, href:, method: } — one link out, e.g. an undo. A
  #             partial one is a mistake worth naming rather than a link that
  #             silently does not render.
  def toast(message, kind: :info, title: nil, duration: nil, action: nil)
    kinds = %i[ info success warning error ]
    unless kinds.include?(kind.to_sym)
      raise ArgumentError, "unknown toast kind #{kind.inspect}, expected one of #{kinds.join(", ")}"
    end

    if action && !(action[:label] && action[:href])
      raise ArgumentError, "a toast action needs both :label and :href, got #{action.inspect}"
    end

    turbo_stream_action_tag :toast, target: "toasts",
                            kind: kind, title: title, duration: duration,
                            "action-label": action&.dig(:label),
                            "action-href": action&.dig(:href),
                            "action-method": action&.dig(:method),
                            template: ERB::Util.html_escape(message)
  end
end
